### PCR Test Sonucunu Yapay Sinir Ağları İLe Test Etme

Problem: Pcr tets sonucu verilen 14 kişinin 7 farklı hastalık belirtileri var ya da yok olarak verilmiştir. Gizli katmansız yapay sinir ağı kurarak yeni hastalık belirtileri bilinen kişilerin Pcr test sonucunu tahmin edelim.

6 input, 1 output ve gizli katmansız yapay sinir ağları yapısında ağırlıklar 0.1, 0.1, 0.1, 0.3, 0.15,0.25 olarak girilmiştir.

In [4]:
import numpy as np
input_ozellikleri = np.array([[1,1,1,1,0,1],[1,0,1,1,1,1],[1,1,0,1,1,1],
                            [0,1,0,1,0,1],[0,0,0,0,1,1],[0,0,1,0,0,0],
                            [1,1,0,1,1,1],[1,0,0,0,0,1],[0,1,0,0,0,1],
                            [0,1,1,1,1,1],[0,0,1,1,1,1],[1,1,0,0,0,1],
                            [1,1,1,1,1,1],[1,0,0,1,0,0]])

In [5]:
print(input_ozellikleri.shape)
input_ozellikleri

(14, 6)


array([[1, 1, 1, 1, 0, 1],
       [1, 0, 1, 1, 1, 1],
       [1, 1, 0, 1, 1, 1],
       [0, 1, 0, 1, 0, 1],
       [0, 0, 0, 0, 1, 1],
       [0, 0, 1, 0, 0, 0],
       [1, 1, 0, 1, 1, 1],
       [1, 0, 0, 0, 0, 1],
       [0, 1, 0, 0, 0, 1],
       [0, 1, 1, 1, 1, 1],
       [0, 0, 1, 1, 1, 1],
       [1, 1, 0, 0, 0, 1],
       [1, 1, 1, 1, 1, 1],
       [1, 0, 0, 1, 0, 0]])

In [17]:
# Hedef Output Tanimlama
hedef_output = np.array([[1,1,1,0,0,0,1,0,0,1,1,0,1,0]])

In [20]:
# Hedef Degişkeni Vektore Donusturme
import numpy as np
hedef_output = hedef_output.reshape(14,1)
print(hedef_output.shape)
print(hedef_output)

(14, 1)
[[1]
 [1]
 [1]
 [0]
 [0]
 [0]
 [1]
 [0]
 [0]
 [1]
 [1]
 [0]
 [1]
 [0]]


In [27]:
# Agirliklari Tanimlama
agirliklar = np.array([[0.1],[0.1],[0.1],[0.3],[0.15],[0.25]])
print(agirliklar.shape)
print(agirliklar)

(6, 1)
[[0.1 ]
 [0.1 ]
 [0.1 ]
 [0.3 ]
 [0.15]
 [0.25]]


In [28]:
# Sigmoid Fonk.
def  sigmoid(x):
    return 1/(1+np.exp(-x))
# Sigmoid fonk. türevi
def sigmoid_der(x):
    return sigmoid(x)*(1-sigmoid(x))

In [29]:
# Yanlilik Agirligi
yanlilik = 0.3
#Ogrenme Orani
lr = 0.05

In [32]:
# Sinir Aglari Yapisi
# Kodları 10000 defa calistirma
for epoch in range(10000):
    inputlar = input_ozellikleri
    
    #İleri Besleme input
    tahmin_input = np.dot(inputlar, agirliklar) + yanlilik
    
    #output
    tahmin_output = sigmoid(tahmin_input)
    
    #Geri yayilim
    #Hatayi hesaplama
    hata = tahmin_output - hedef_output
    x = hata.sum()
    print(x)

   #Turevi Hesaplama
    dcost_dpred = hata
    dpred_dz = sigmoid_der(tahmin_output)
    
    # Kismi Turevleri Hesaplama
    z_delta = dcost_dpred * dpred_dz
    
    # 3.dereceden kısmi turevleri hesaplama
    inputlar = input_ozellikleri.T
    agirliklar -= lr * np.dot(inputlar, z_delta)
    
    # Yanlilik Agirlik Degeri Guncelleme
    for i in z_delta:
        yanlilik -= lr * i



0.005147399719275656
0.005147139191528971
0.005146878690125006
0.005146618215059311
0.0051463577663287945
0.00514609734392857
0.005145836947854959
0.00514557657810443
0.005145316234672147
0.005145055917554627
0.005144795626747846
0.005144535362248117
0.005144275124050947
0.005144014912152166
0.005143754726549189
0.0051434945672363085
0.005143234434210369
0.0051429743274676466
0.005142714247003293
0.005142454192814068
0.005142194164895688
0.005141934163243718
0.005141674187855396
0.005141414238725996
0.005141154315850908
0.005140894419227958
0.0051406345488516515
0.005140374704718566
0.005140114886824573
0.005139855095165064
0.005139595329736565
0.005139335590536049
0.005139075877558581
0.005138816190800104
0.005138556530257688
0.005138296895925563
0.0051380372878015845
0.005137777705880573
0.0051375181501588435
0.005137258620632882
0.005136999117298724
0.005136739640151902
0.005136480189188907
0.005136220764404071
0.005135961365796636
0.00513570199335981
0.005135442647091617
0.00513518

0.0048693672262502125
0.004869134054803615
0.004868900905663364
0.004868667778827908
0.004868434674292571
0.004868201592054257
0.004867968532110347
0.004867735494457126
0.004867502479091035
0.004867269486009754
0.00486703651520958
0.004866803566688163
0.004866570640440972
0.004866337736465191
0.0048661048547588084
0.004865871995317377
0.004865639158137724
0.004865406343217271
0.004865173550552911
0.0048649407801401945
0.00486470803197738
0.00486447530606123
0.004864242602386955
0.004864009920952545
0.004863777261755358
0.004863544624791568
0.004863312010058236
0.004863079417551304
0.004862846847268442
0.004862614299206014
0.004862381773362177
0.0048621492697315935
0.004861916788312258
0.004861684329101374
0.004861451892094106
0.004861219477289834
0.004860987084683174
0.004860754714271905
0.004860522366053006
0.004860290040022775
0.00486005773617838
0.004859825454516161
0.004859593195032987
0.004859360957726719
0.004859128742593083
0.0048588965496292776
0.004858664378831855
0.0048584322

0.004667737075261814
0.004667522795396034
0.0046673085351830665
0.004667094294619903
0.004666880073704735
0.0046666658724352494
0.004666451690807198
0.00466623752881958
0.0046660233864681214
0.004665809263750927
0.004665595160664971
0.0046653810772078926
0.004665167013377115
0.004664952969169274
0.004664738944582813
0.004664524939613754
0.004664310954260112
0.004664096988518944
0.00466388304238803
0.0046636691158640714
0.004663455208944569
0.004663241321627062
0.004663027453908477
0.004662813605786028
0.0046625997772580564
0.004662385968320192
0.0046621721789717385
0.0046619584092088195
0.0046617446590283
0.0046615309284276304
0.004661317217405715
0.004661103525958037
0.004660889854082407
0.004660676201776886
0.004660462569037795
0.0046602489558629934
0.004660035362249416
0.004659821788194723
0.004659608233696463
0.004659394698751148
0.004659181183356665
0.0046589676875105485
0.004658754211209726
0.004658540754451572
0.004658327317232976
0.004658113899551963
0.00465790050140609
0.00465

0.004449951957653698
0.0044497571880547145
0.004449562435487511
0.004449367699950806
0.004449172981441895
0.004448978279959009
0.004448783595499436
0.004448588928060844
0.004448394277641316
0.004448199644238452
0.004448005027850905
0.004447810428474804
0.00444761584610912
0.0044474212807512695
0.004447226732398821
0.004447032201049805
0.004446837686701847
0.00444664318935282
0.004446448709000925
0.004446254245642737
0.004446059799277629
0.004445865369902565
0.004445670957515072
0.0044454765621124255
0.0044452821836943575
0.0044450878222564326
0.004444893477797763
0.004444699150315956
0.004444504839808359
0.004444310546273128
0.004444116269708224
0.004443922010110812
0.004443727767478818
0.004443533541810245
0.00444333933310254
0.004443145141354597
0.004442950966562417
0.004442756808725409
0.0044425626678401085
0.004442368543905381
0.004442174436917867
0.004441980346876531
0.004441786273778421
0.004441592217621103
0.004441398178402821
0.0044412041561212555
0.004441010150774127
0.0044408

0.0042421630152208855
0.004241985993763535
0.004241808987065537
0.00424163199512547
0.004241455017941292
0.004241278055509976
0.004241101107831417
0.004240924174902072
0.004240747256722048
0.004240570353287742
0.0042403934645979
0.004240216590650358
0.004240039731444902
0.004239862886978274
0.004239686057248029
0.004239509242253419
0.004239332441992893
0.004239155656462917
0.004238978885663847
0.004238802129592753
0.004238625388246969
0.004238448661626464
0.004238271949727419
0.004238095252549565
0.004237918570090399
0.004237741902348678
0.0042375652493210365
0.004237388611007467
0.004237211987405107
0.004237035378512518
0.004236858784327849
0.004236682204848739
0.00423650564007398
0.004236329090001139
0.004236152554629178
0.004235976033955776
0.004235799527978477
0.004235623036696829
0.004235446560108049
0.0042352700982106195
0.004235093651002412
0.004234917218481807
0.004234740800647192
0.004234564397496055
0.004234388009027467
0.004234211635239157
0.004234035276128799
0.004233858931

0.004091396881628145
0.004091232207858539
0.004091067547332589
0.004090902900048192
0.004090738266002411
0.004090573645196062
0.004090409037626119
0.00409024444329107
0.004090079862189641
0.0040899152943202665
0.004089750739681193
0.004089586198270703
0.0040894216700869245
0.004089257155128944
0.00408909265339445
0.004088928164883446
0.004088763689591896
0.004088599227520161
0.004088434778666045
0.004088270343027216
0.004088105920603552
0.004087941511391873
0.004087777115391988
0.004087612732601579
0.004087448363018931
0.004087284006642393
0.004087119663471285
0.004086955333502899
0.004086791016736199
0.004086626713169653
0.004086462422801505
0.004086298145630396
0.004086133881654602
0.004085969630872217
0.004085805393282212
0.004085641168882588
0.004085476957671844
0.004085312759648433
0.0040851485748106065
0.004084984403157144
0.004084820244686293
0.004084656099396851
0.004084491967285893
0.004084327848353404
0.0040841637425968105
0.004083999650014575
0.00408383557060559
0.0040836715

0.003918073209237996
0.003917922179878872
0.0039177711621516875
0.003917620156055625
0.0039174691615898975
0.0039173181787516345
0.003917167207541888
0.003917016247956766
0.003916865299995984
0.003916714363658515
0.003916563438943034
0.003916412525847372
0.003916261624371075
0.0039161107345125155
0.0039159598562699665
0.003915808989642511
0.0039156581346290395
0.003915507291228006
0.003915356459437323
0.003915205639256155
0.003915054830683856
0.003914904033718372
0.003914753248358603
0.003914602474603318
0.003914451712450428
0.0039143009618996406
0.003914150222948864
0.00391399949559699
0.003913848779842862
0.003913698075684291
0.003913547383121474
0.0039133967021518935
0.00391324603277525
0.003913095374988814
0.003912944728791742
0.003912794094183528
0.003912643471161964
0.003912492859725988
0.003912342259874438
0.003912191671605388
0.003912041094917983
0.003911890529811192
0.003911739976282889
0.003911589434332227
0.003911438903957645
0.0039112883851582505
0.003911137877932014
0.0039

0.003763279860545552
0.003763140518794913
0.003763001187353571
0.0037628618662201366
0.0037627225553938633
0.003762583254872943
0.0037624439646571617
0.003762304684745298
0.0037621654151350613
0.0037620261558264814
0.003761886906817666
0.0037617476681076826
0.00376160843969554
0.0037614692215797657
0.003761330013760333
0.003761190816234272
0.003761051629001683
0.003760912452061712
0.003760773285412072
0.003760634129052842
0.0037604949829821615
0.0037603558471990496
0.0037602167217017815
0.003760077606490004
0.0037599385015622882
0.0037597994069177177
0.003759660322554916
0.003759521248472738
0.0037593821846701888
0.0037592431311459966
0.0037591040878987843
0.0037589650549281154
0.003758826032232243
0.0037586870198098197
0.0037585480176604478
0.0037584090257829113
0.0037582700441756546
0.003758131072837423
0.00375799211176727
0.0037578531609645293
0.003757714220427766
0.003757575290156
0.0037574363701469134
0.003757297460400857
0.0037571585609157392
0.0037570196716915144
0.0037568807927

0.003644913073514673
0.0036447823518863474
0.0036446516396259608
0.003644520936732622
0.0036443902432049353
0.003644259559042179
0.0036441288842435898
0.0036439982188078814
0.003643867562733791
0.0036437369160210865
0.003643606278668108
0.0036434756506733878
0.003643345032036776
0.00364321442275758
0.0036430838228334447
0.0036429532322649343
0.0036428226510501578
0.003642692079187963
0.0036425615166772642
0.0036424309635178778
0.003642300419709015
0.0036421698852474463
0.0036420393601352993
0.0036419088443692188
0.0036417783379486452
0.0036416478408732057
0.0036415173531411248
0.0036413868747526574
0.0036412564057054786
0.0036411259459993967
0.003640995495632802
0.003640865054605208
0.0036407346229156714
0.0036406042005625947
0.003640473787545233
0.003640343383863297
0.003640212989514656
0.003640082604498635
0.0036399522288145074
0.0036398218624611573
0.00363969150543748
0.003639561157742933
0.0036394308193757534
0.003639300490335408
0.0036391701706211684
0.0036390398602316612
0.003638

0.003510928848522884
0.0035108075527062106
0.0035106862652622426
0.0035105649861910944
0.003510443715491926
0.0035103224531640603
0.00351020119920613
0.0035100799536175197
0.003509958716397007
0.0035098374875448604
0.003509716267058705
0.0035095950549389027
0.003509473851183554
0.0035093526557922797
0.003509231468764065
0.0035091102900983943
0.003508989119794828
0.0035088679578510976
0.0035087468042670285
0.003508625659042553
0.003508504522175383
0.003508383393666343
0.0035082622735128437
0.0035081411617146584
0.003508020058271453
0.003507898963181758
0.0035077778764448764
0.0035076567980599793
0.0035075357280261246
0.0035074146663421606
0.0035072936130086346
0.003507172568022436
0.0035070515313842376
0.003506930503093802
0.0035068094831480806
0.0035066884715475935
0.003506567468291515
0.0035064464733793282
0.0035063254868091544
0.00350620450858036
0.0035060835386928183
0.0035059625771441325
0.003505841623935618
0.003505720679065475
0.0035055997425315188
0.0035054788143349656
0.0035053

In [33]:
# Tahmin Agirliklari ve Yanliligi Yazdirma
print(agirliklar)

[[3.58323349]
 [3.5387087 ]
 [7.24781575]
 [4.32894841]
 [7.63344067]
 [0.09366125]]


In [34]:
print(yanlilik)

[-13.37216901]


Model kurulduktan sonra yeni özellik değerleri girilen PCR test sonuçları tahmin edilebilir.

In [35]:
# Yeni inputlar girme
yeni_ozellikler = np.array([1,0,0,1,1,1])

# İlk adim
tahmin1 = np.dot(yeni_ozellikler, agirliklar) + yanlilik

# İkinci adim
tahmin2 = sigmoid(tahmin1)

#Sonuc tahmini yazdirma
print(tahmin2)

[0.90611663]


Yorum: Olasılık 0.9061 > 0.5 olduğundan tahmin 1  olarak yani pozitif olarak sınıflandırılır.

In [37]:
# Yeni inputlar girme
yeni_ozellikler = np.array([0,0,0,1,1,1])

# İlk adim
tahmin1 = np.dot(yeni_ozellikler, agirliklar) + yanlilik

# İkinci adim
tahmin2 = sigmoid(tahmin1)

#Sonuc tahmini yazdirma
print(tahmin2)

[0.21146477]


Yorum: Olasılık değeri 0.2114 < 0.5 olduğundan tahmin 0 olarak yani negatif olarak sınıflandırılır.

In [39]:
# Yeni inputlar girme
yeni_ozellikler = np.array([1,0,1,1,1,1])

# İlk adim
tahmin1 = np.dot(yeni_ozellikler, agirliklar) + yanlilik

# İkinci adim
tahmin2 = sigmoid(tahmin1)

#Sonuc tahmini yazdirma
print(tahmin2)

[0.99992626]


Yorum: Olasılık değeri 0.9999 > 0.5 olduğundan tahmin 1 olarak yani pozitif olarak sınıflandırılır.